File: leetcode_py/p1031.py
from typing import List
from itertools import *

# class Solution:
#     def maxSumTwoNoOverlap(self, nums: List[int], firstLen: int, secondLen: int) -> int:
#         #  双指针
#         # 重叠判断
#         # i - j   j 比 i 大的时候 就应该 直接 遍历
#         # 预出列
#         n = len(nums)
#         preSum = [0] * (n + 1)
#         for i in range(1, n + 1):
#             preSum[i] = preSum[i - 1] + nums[i - 1]
#         print(preSum)
#         # 处理值
#         res = 0
#         for i in range(firstLen, n + 1):
#             for j in range(secondLen, n + 1):
#                 if secondLen <= j - i or firstLen <= i - j:
#                     print(i, j, res)
#                     res = max(res, preSum[i] - preSum[i -
#                                                       firstLen] + preSum[j] - preSum[j - secondLen])

#         return res


class Solution:
    def maxSumTwoNoOverlap(self, nums: List[int], firstLen: int, secondLen: int) -> int:
        s = list(accumulate(nums, initial=0))
        ans = 0
        maxSumA = maxSumB = 0

        for i in range(firstLen + secondLen, len(s)):
            # 维护最大的 左边的 a
            maxSumA = max(maxSumA, s[i - secondLen] -
                          s[i - secondLen - firstLen])
            maxSumB = max(maxSumB, s[i - firstLen] -
                          s[i - secondLen - firstLen])
            # 更新 目前的 a + b
            ans = max(ans, maxSumA + s[i] - s[i - secondLen],
                      maxSumB + s[i] - s[i - firstLen])

        return ans

File: leetcode_C++/p1031.cpp
/*
@Author: mrxhar
@Date:   2023-09-13 09:40:45
@Email:  1337904145@qq.com
@Blog:   https://leetcode.cn/problems/maximum-sum-of-two-non-overlapping-subarrays/description/?envType=daily-question&envId=2023-09-13
*/
#include <vector>

using namespace std;

class Solution {
public:
    int maxSumTwoNoOverlap(vector<int>& nums, int firstLen, int secondLen) {
        // 固定一个枚举第二个，优化第一个变量
        int n = nums.size();
        vector<int> sum;
        sum.push_back(0);
        for (auto& num : nums) {
            sum.push_back(sum.back() + num);
        }

        int ans = 0;
        int maxA = 0, maxB = 0;
        for (int i = firstLen + secondLen; i < sum.size(); i++) {
            maxA = max(maxA, sum[i - secondLen] - sum[i - secondLen - firstLen]);
            maxB = max(maxB, sum[i - firstLen] - sum[i - secondLen - firstLen]);
            ans = max(ans, max(maxA + sum[i] - sum[i - secondLen], maxB + sum[i] - sum[i - firstLen]));
        }
        return ans;
    }
};